Property 'onProgress' does not exist on type 'Scraper<ScraperCredentials>'

In https://github.com/eshaham/israeli-bank-scrapers/pull/760 the signature of  `createScraper` changed -
```diff
- export default function createScraper(options: ScraperOptions) {
+ export default function createScraper(options: ScraperOptions): Scraper<ScraperCredentials> {
```

```ts
export interface Scraper<TCredentials extends ScraperCredentials> {
  scrape(credentials: TCredentials): Promise<ScraperScrapingResult>;
}
```

Since the new interface has only the `scrape` method, the following error is emitted when accessing methods of `BaseScraper` like `onProgress`:
```
error TS2339: Property 'onProgress' does not exist on type 'Scraper<ScraperCredentials>'.
     scraper.onProgress((companyId, { type }) => {
```

--- a/src/definitions.ts
+++ b/src/definitions.ts
@@ -92,3 +92,14 @@ export const SCRAPERS = {
     loginFields: ['email', PASSWORD_FIELD, 'otpCodeRetriever', 'phoneNumber', 'otpLongTermToken'],
   },
 };
+
+export enum ScraperProgressTypes {
+  Initializing = 'INITIALIZING',
+  StartScraping = 'START_SCRAPING',
+  LoggingIn = 'LOGGING_IN',
+  LoginSuccess = 'LOGIN_SUCCESS',
+  LoginFailed = 'LOGIN_FAILED',
+  ChangePassword = 'CHANGE_PASSWORD',
+  EndScraping = 'END_SCRAPING',
+  Terminating = 'TERMINATING',
+}

--- a/src/scrapers/base-isracard-amex.ts
+++ b/src/scrapers/base-isracard-amex.ts
@@ -16,14 +16,11 @@ import {
   TransactionsAccount, Transaction, TransactionInstallments,
   TransactionStatuses, TransactionTypes,
 } from '../transactions';
-import {
-  ScraperProgressTypes,
-
-} from './base-scraper';
 import { getDebug } from '../helpers/debug';
 import { runSerial } from '../helpers/waiting';
 import { ScraperErrorTypes } from './errors';
 import { ScraperScrapingResult, ScraperOptions } from './interface';
+import { ScraperProgressTypes } from '../definitions';
 
 const COUNTRY_CODE = '212';
 const ID_TYPE = '1';

--- a/src/scrapers/base-scraper-with-browser.ts
+++ b/src/scrapers/base-scraper-with-browser.ts
@@ -77,31 +77,6 @@ async function getKeyByValue(object: PossibleLoginResults, value: string, page: 
   return Promise.resolve(LoginResults.UnknownError);
 }
 
-function handleLoginResult(scraper: BaseScraperWithBrowser<ScraperCredentials>, loginResult: LoginResults) {
-  switch (loginResult) {
-    case LoginResults.Success:
-      scraper.emitProgress(ScraperProgressTypes.LoginSuccess);
-      return { success: true };
-    case LoginResults.InvalidPassword:
-    case LoginResults.UnknownError:
-      scraper.emitProgress(ScraperProgressTypes.LoginFailed);
-      return {
-        success: false,
-        errorType: loginResult === LoginResults.InvalidPassword ? ScraperErrorTypes.InvalidPassword :
-          ScraperErrorTypes.General,
-        errorMessage: `Login failed with ${loginResult} error`,
-      };
-    case LoginResults.ChangePassword:
-      scraper.emitProgress(ScraperProgressTypes.ChangePassword);
-      return {
-        success: false,
-        errorType: ScraperErrorTypes.ChangePassword,
-      };
-    default:
-      throw new Error(`unexpected login result "${loginResult}"`);
-  }
-}
-
 function createGeneralError(): ScraperScrapingResult {
   return {
     success: false,
@@ -110,11 +85,11 @@ function createGeneralError(): ScraperScrapingResult {
 }
 
 class BaseScraperWithBrowser<TCredentials extends ScraperCredentials> extends BaseScraper<TCredentials> {
-  // NOTICE - it is discourage to use bang (!) in general. It is used here because
+  // NOTICE - it is discouraged to use bang (!) in general. It is used here because
   // all the classes that inherit from this base assume is it mandatory.
   protected browser!: Browser;
 
-  // NOTICE - it is discourage to use bang (!) in general. It is used here because
+  // NOTICE - it is discouraged to use bang (!) in general. It is used here because
   // all the classes that inherit from this base assume is it mandatory.
   protected page!: Page;
 
@@ -279,7 +254,7 @@ class BaseScraperWithBrowser<TCredentials extends ScraperCredentials> extends Ba
     const current = await getCurrentUrl(this.page, true);
     const loginResult = await getKeyByValue(loginOptions.possibleResults, current, this.page);
     debug(`handle login results ${loginResult}`);
-    return handleLoginResult(this, loginResult);
+    return this.handleLoginResult(loginResult);
   }
 
   async terminate(_success: boolean) {
@@ -299,6 +274,31 @@ class BaseScraperWithBrowser<TCredentials extends ScraperCredentials> extends Ba
     }
 
     await this.browser.close();
+  }
+
+  private handleLoginResult(loginResult: LoginResults) {
+    switch (loginResult) {
+      case LoginResults.Success:
+        this.emitProgress(ScraperProgressTypes.LoginSuccess);
+        return { success: true };
+      case LoginResults.InvalidPassword:
+      case LoginResults.UnknownError:
+        this.emitProgress(ScraperProgressTypes.LoginFailed);
+        return {
+          success: false,
+          errorType: loginResult === LoginResults.InvalidPassword ? ScraperErrorTypes.InvalidPassword :
+            ScraperErrorTypes.General,
+          errorMessage: `Login failed with ${loginResult} error`,
+        };
+      case LoginResults.ChangePassword:
+        this.emitProgress(ScraperProgressTypes.ChangePassword);
+        return {
+          success: false,
+          errorType: ScraperErrorTypes.ChangePassword,
+        };
+      default:
+        throw new Error(`unexpected login result "${loginResult}"`);
+    }
   }
 }
 

--- a/src/scrapers/base-scraper-with-browser.ts
+++ b/src/scrapers/base-scraper-with-browser.ts
@@ -1,7 +1,7 @@
 import puppeteer, { Browser, Frame, Page } from 'puppeteer';
 
 import {
-  BaseScraper, ScraperProgressTypes,
+  BaseScraper,
 
 } from './base-scraper';
 import { getCurrentUrl, waitForNavigation } from '../helpers/navigation';
@@ -9,6 +9,7 @@ import { clickButton, fillInput, waitUntilElementFound } from '../helpers/elemen
 import { getDebug } from '../helpers/debug';
 import { ScraperErrorTypes } from './errors';
 import { ScraperScrapingResult, ScraperCredentials } from './interface';
+import { ScraperProgressTypes } from '../definitions';
 
 const VIEWPORT_WIDTH = 1024;
 const VIEWPORT_HEIGHT = 768;

--- a/src/scrapers/base-scraper.ts
+++ b/src/scrapers/base-scraper.ts
@@ -77,17 +77,17 @@ export class BaseScraper<TCredentials extends ScraperCredentials> implements Scr
   }
 
   // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/require-await
-  async login(_credentials: TCredentials): Promise<ScraperLoginResult> {
+  protected async login(_credentials: TCredentials): Promise<ScraperLoginResult> {
     throw new Error(`login() is not created in ${this.options.companyId}`);
   }
 
   // eslint-disable-next-line  @typescript-eslint/require-await
-  async fetchData(): Promise<ScraperScrapingResult> {
+  protected async fetchData(): Promise<ScraperScrapingResult> {
     throw new Error(`fetchData() is not created in ${this.options.companyId}`);
   }
 
   // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/require-await
-  async terminate(_success: boolean) {
+  protected async terminate(_success: boolean) {
     this.emitProgress(ScraperProgressTypes.Terminating);
   }
 

--- a/src/scrapers/base-scraper.ts
+++ b/src/scrapers/base-scraper.ts
@@ -91,11 +91,11 @@ export class BaseScraper<TCredentials extends ScraperCredentials> implements Scr
     this.emitProgress(ScraperProgressTypes.Terminating);
   }
 
-  emitProgress(type: ScraperProgressTypes) {
+  protected emitProgress(type: ScraperProgressTypes) {
     this.emit(SCRAPE_PROGRESS, { type });
   }
 
-  emit(eventName: string, payload: Record<string, any>) {
+  protected emit(eventName: string, payload: Record<string, any>) {
     this.eventEmitter.emit(eventName, this.options.companyId, payload);
   }
 

--- a/src/scrapers/base-scraper.ts
+++ b/src/scrapers/base-scraper.ts
@@ -11,20 +11,10 @@ import {
   ScraperScrapingResult,
   ScraperTwoFactorAuthTriggerResult,
 } from './interface';
+import { CompanyTypes, ScraperProgressTypes } from '../definitions';
 
 const SCRAPE_PROGRESS = 'SCRAPE_PROGRESS';
 
-
-export enum ScraperProgressTypes {
-  Initializing = 'INITIALIZING',
-  StartScraping = 'START_SCRAPING',
-  LoggingIn = 'LOGGING_IN',
-  LoginSuccess = 'LOGIN_SUCCESS',
-  LoginFailed = 'LOGIN_FAILED',
-  ChangePassword = 'CHANGE_PASSWORD',
-  EndScraping = 'END_SCRAPING',
-  Terminating = 'TERMINATING',
-}
 
 export class BaseScraper<TCredentials extends ScraperCredentials> implements Scraper<TCredentials> {
   private eventEmitter = new EventEmitter();
@@ -109,7 +99,7 @@ export class BaseScraper<TCredentials extends ScraperCredentials> implements Scr
     this.eventEmitter.emit(eventName, this.options.companyId, payload);
   }
 
-  onProgress(func: (...args: any[]) => void) {
+  onProgress(func: (companyId: CompanyTypes, payload: {type: ScraperProgressTypes}) => void) {
     this.eventEmitter.on(SCRAPE_PROGRESS, func);
   }
 }

--- a/src/scrapers/factory.test.ts
+++ b/src/scrapers/factory.test.ts
@@ -1,0 +1,16 @@
+/* eslint-disable @typescript-eslint/unbound-method */
+import { CompanyTypes } from '../definitions';
+import createScraper from './factory';
+
+describe('Factory', () => {
+  test('should return a scraper instance', () => {
+    const scraper = createScraper({
+      companyId: CompanyTypes.hapoalim,
+      startDate: new Date(),
+    });
+    expect(scraper).toBeDefined();
+
+    expect(scraper.scrape).toBeInstanceOf(Function);
+    expect(scraper.onProgress).toBeInstanceOf(Function);
+  });
+});

--- a/src/scrapers/interface.ts
+++ b/src/scrapers/interface.ts
@@ -138,6 +138,8 @@ export interface ScraperScrapingResult {
 export interface Scraper<TCredentials extends ScraperCredentials> {
   scrape(credentials: TCredentials): Promise<ScraperScrapingResult>;
   onProgress(func: (companyId: CompanyTypes, payload: {type: ScraperProgressTypes}) => void): void;
+  triggerTwoFactorAuth(phoneNumber: string): Promise<ScraperTwoFactorAuthTriggerResult>;
+  getLongTermTwoFactorToken(otpCode: string): Promise<ScraperGetLongTermTwoFactorTokenResult>;
 }
 
 export type ScraperTwoFactorAuthTriggerResult = ErrorResult | {

--- a/src/scrapers/interface.ts
+++ b/src/scrapers/interface.ts
@@ -1,5 +1,5 @@
 import { Browser, Page } from 'puppeteer';
-import { CompanyTypes } from '../definitions';
+import { CompanyTypes, ScraperProgressTypes } from '../definitions';
 import { TransactionsAccount } from '../transactions';
 import { ErrorResult, ScraperErrorTypes } from './errors';
 
@@ -137,6 +137,7 @@ export interface ScraperScrapingResult {
 
 export interface Scraper<TCredentials extends ScraperCredentials> {
   scrape(credentials: TCredentials): Promise<ScraperScrapingResult>;
+  onProgress(func: (companyId: CompanyTypes, payload: {type: ScraperProgressTypes}) => void): void;
 }
 
 export type ScraperTwoFactorAuthTriggerResult = ErrorResult | {

